// kysely-codegen/dist/dialects/postgres/postgres-adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostgresAdapter = void 0;
const adapter_1 = require("../../adapter");
const constants_1 = require("../../constants");
const generic_expression_node_1 = require("../../nodes/generic-expression-node");
const identifier_node_1 = require("../../nodes/identifier-node");
const object_expression_node_1 = require("../../nodes/object-expression-node");
const property_node_1 = require("../../nodes/property-node");
const union_expression_node_1 = require("../../nodes/union-expression-node");
class PostgresAdapter extends adapter_1.Adapter {
    constructor() {
        super(...arguments);
        // From https://node-postgres.com/features/types:
        // "node-postgres will convert a database type to a JavaScript string if it doesn't have a
        // registered type parser for the database type. Furthermore, you can send any type to the
        // PostgreSQL server as a string and node-postgres will pass it through without modifying it in
        // any way."
        this.defaultType = new identifier_node_1.IdentifierNode('string');
        this.definitions = {
            Circle: new object_expression_node_1.ObjectExpressionNode([
                new property_node_1.PropertyNode('x', new identifier_node_1.IdentifierNode('number')),
                new property_node_1.PropertyNode('y', new identifier_node_1.IdentifierNode('number')),
                new property_node_1.PropertyNode('radius', new identifier_node_1.IdentifierNode('number')),
            ]),
            Int8: generic_expression_node_1.GenericExpressionNode.createColumnType(new identifier_node_1.IdentifierNode('string'), new union_expression_node_1.UnionExpressionNode([
                new identifier_node_1.IdentifierNode('string'),
                new identifier_node_1.IdentifierNode('number'),
                new identifier_node_1.IdentifierNode('bigint'),
            ])),
            Interval: generic_expression_node_1.GenericExpressionNode.createColumnType(new identifier_node_1.IdentifierNode('IPostgresInterval'), new union_expression_node_1.UnionExpressionNode([
                new identifier_node_1.IdentifierNode('IPostgresInterval'),
                new identifier_node_1.IdentifierNode('number'),
            ])),
            Json: generic_expression_node_1.GenericExpressionNode.createColumnType(new identifier_node_1.IdentifierNode('JsonValue'), new identifier_node_1.IdentifierNode('string'), new identifier_node_1.IdentifierNode('string')),
            JsonArray: constants_1.JSON_ARRAY_DEFINITION,
            JsonObject: constants_1.JSON_OBJECT_DEFINITION,
            JsonPrimitive: constants_1.JSON_PRIMITIVE_DEFINITION,
            JsonValue: constants_1.JSON_VALUE_DEFINITION,
            Numeric: generic_expression_node_1.GenericExpressionNode.createColumnType(new identifier_node_1.IdentifierNode('string'), new union_expression_node_1.UnionExpressionNode([
                new identifier_node_1.IdentifierNode('string'),
                new identifier_node_1.IdentifierNode('number'),
            ])),
            Point: new object_expression_node_1.ObjectExpressionNode([
                new property_node_1.PropertyNode('x', new identifier_node_1.IdentifierNode('number')),
                new property_node_1.PropertyNode('y', new identifier_node_1.IdentifierNode('number')),
            ]),
            Timestamp: generic_expression_node_1.GenericExpressionNode.createColumnType(new union_expression_node_1.UnionExpressionNode([
                new identifier_node_1.IdentifierNode('Date'),
                new identifier_node_1.IdentifierNode('RawBuilder'),
            ]), new union_expression_node_1.UnionExpressionNode([
                new identifier_node_1.IdentifierNode('Date'),
                new identifier_node_1.IdentifierNode('string'),
                new identifier_node_1.IdentifierNode('RawBuilder'),
            ])),
        };
        this.imports = {
            IPostgresInterval: 'postgres-interval',
        };
        // These types have been found through experimentation in Adminer and in the 'pg' source code.
        this.types = {
            bit: new identifier_node_1.IdentifierNode('string'),
            bool: new identifier_node_1.IdentifierNode('boolean'),
            box: new identifier_node_1.IdentifierNode('string'),
            bpchar: new identifier_node_1.IdentifierNode('string'),
            bytea: new identifier_node_1.IdentifierNode('Buffer'),
            cidr: new identifier_node_1.IdentifierNode('string'),
            circle: new identifier_node_1.IdentifierNode('Circle'),
            date: new identifier_node_1.IdentifierNode('Timestamp'),
            float4: new identifier_node_1.IdentifierNode('number'),
            float8: new identifier_node_1.IdentifierNode('number'),
            inet: new identifier_node_1.IdentifierNode('string'),
            int2: new identifier_node_1.IdentifierNode('number'),
            int4: new identifier_node_1.IdentifierNode('number'),
            int8: new identifier_node_1.IdentifierNode('Int8'),
            interval: new identifier_node_1.IdentifierNode('Interval'),
            json: new identifier_node_1.IdentifierNode('Json'),
            jsonb: new identifier_node_1.IdentifierNode('Json'),
            line: new identifier_node_1.IdentifierNode('string'),
            lseg: new identifier_node_1.IdentifierNode('string'),
            macaddr: new identifier_node_1.IdentifierNode('string'),
            money: new identifier_node_1.IdentifierNode('string'),
            numeric: new identifier_node_1.IdentifierNode('Numeric'),
            oid: new identifier_node_1.IdentifierNode('number'),
            path: new identifier_node_1.IdentifierNode('string'),
            point: new identifier_node_1.IdentifierNode('Point'),
            polygon: new identifier_node_1.IdentifierNode('string'),
            text: new identifier_node_1.IdentifierNode('string'),
            time: new identifier_node_1.IdentifierNode('string'),
            timestamp: new identifier_node_1.IdentifierNode('Timestamp'),
            timestamptz: new identifier_node_1.IdentifierNode('Timestamp'),
            tsquery: new identifier_node_1.IdentifierNode('string'),
            tsvector: new identifier_node_1.IdentifierNode('string'),
            txid_snapshot: new identifier_node_1.IdentifierNode('string'),
            uuid: new identifier_node_1.IdentifierNode('string'),
            varbit: new identifier_node_1.IdentifierNode('string'),
            varchar: new identifier_node_1.IdentifierNode('string'),
            xml: new identifier_node_1.IdentifierNode('string'),
        };
    }
}
exports.PostgresAdapter = PostgresAdapter;
