// kysely-codegen/dist/dialects/mysql/mysql-adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MysqlAdapter = void 0;
const adapter_1 = require("../../adapter");
const constants_1 = require("../../constants");
const nodes_1 = require("../../nodes");
const array_expression_node_1 = require("../../nodes/array-expression-node");
const identifier_node_1 = require("../../nodes/identifier-node");
const object_expression_node_1 = require("../../nodes/object-expression-node");
const property_node_1 = require("../../nodes/property-node");
const union_expression_node_1 = require("../../nodes/union-expression-node");
class MysqlAdapter extends adapter_1.Adapter {
    constructor() {
        super(...arguments);
        this.definitions = {
            Decimal: nodes_1.GenericExpressionNode.createColumnType(new identifier_node_1.IdentifierNode('string'), new union_expression_node_1.UnionExpressionNode([
                new identifier_node_1.IdentifierNode('string'),
                new identifier_node_1.IdentifierNode('number'),
            ])),
            Geometry: new union_expression_node_1.UnionExpressionNode([
                new identifier_node_1.IdentifierNode('LineString'),
                new identifier_node_1.IdentifierNode('Point'),
                new identifier_node_1.IdentifierNode('Polygon'),
                new array_expression_node_1.ArrayExpressionNode(new identifier_node_1.IdentifierNode('Geometry')),
            ]),
            Json: nodes_1.GenericExpressionNode.createColumnType(new identifier_node_1.IdentifierNode('JsonValue'), new identifier_node_1.IdentifierNode('string'), new identifier_node_1.IdentifierNode('string')),
            JsonArray: constants_1.JSON_ARRAY_DEFINITION,
            JsonObject: constants_1.JSON_OBJECT_DEFINITION,
            JsonPrimitive: constants_1.JSON_PRIMITIVE_DEFINITION,
            JsonValue: constants_1.JSON_VALUE_DEFINITION,
            LineString: new array_expression_node_1.ArrayExpressionNode(new identifier_node_1.IdentifierNode('Point')),
            Point: new object_expression_node_1.ObjectExpressionNode([
                new property_node_1.PropertyNode('x', new identifier_node_1.IdentifierNode('number')),
                new property_node_1.PropertyNode('y', new identifier_node_1.IdentifierNode('number')),
            ]),
            Polygon: new array_expression_node_1.ArrayExpressionNode(new identifier_node_1.IdentifierNode('LineString')),
        };
        // These types have been found through experimentation in Adminer.
        this.types = {
            bigint: new identifier_node_1.IdentifierNode('number'),
            binary: new identifier_node_1.IdentifierNode('Buffer'),
            bit: new identifier_node_1.IdentifierNode('Buffer'),
            blob: new identifier_node_1.IdentifierNode('Buffer'),
            char: new identifier_node_1.IdentifierNode('string'),
            date: new identifier_node_1.IdentifierNode('Date'),
            datetime: new identifier_node_1.IdentifierNode('Date'),
            decimal: new identifier_node_1.IdentifierNode('Decimal'),
            double: new identifier_node_1.IdentifierNode('number'),
            enum: new identifier_node_1.IdentifierNode('unknown'),
            float: new identifier_node_1.IdentifierNode('number'),
            geomcollection: new array_expression_node_1.ArrayExpressionNode(new identifier_node_1.IdentifierNode('Geometry')),
            geometry: new identifier_node_1.IdentifierNode('Geometry'),
            int: new identifier_node_1.IdentifierNode('number'),
            json: new identifier_node_1.IdentifierNode('Json'),
            linestring: new identifier_node_1.IdentifierNode('LineString'),
            longblob: new identifier_node_1.IdentifierNode('Buffer'),
            longtext: new identifier_node_1.IdentifierNode('string'),
            mediumblob: new identifier_node_1.IdentifierNode('Buffer'),
            mediumint: new identifier_node_1.IdentifierNode('number'),
            mediumtext: new identifier_node_1.IdentifierNode('string'),
            multilinestring: new array_expression_node_1.ArrayExpressionNode(new identifier_node_1.IdentifierNode('LineString')),
            multipoint: new array_expression_node_1.ArrayExpressionNode(new identifier_node_1.IdentifierNode('Point')),
            multipolygon: new array_expression_node_1.ArrayExpressionNode(new identifier_node_1.IdentifierNode('Polygon')),
            point: new identifier_node_1.IdentifierNode('Point'),
            polygon: new identifier_node_1.IdentifierNode('Polygon'),
            set: new identifier_node_1.IdentifierNode('unknown'),
            smallint: new identifier_node_1.IdentifierNode('number'),
            text: new identifier_node_1.IdentifierNode('string'),
            time: new identifier_node_1.IdentifierNode('string'),
            timestamp: new identifier_node_1.IdentifierNode('Date'),
            tinyblob: new identifier_node_1.IdentifierNode('Buffer'),
            tinyint: new identifier_node_1.IdentifierNode('number'),
            tinytext: new identifier_node_1.IdentifierNode('string'),
            varbinary: new identifier_node_1.IdentifierNode('Buffer'),
            varchar: new identifier_node_1.IdentifierNode('string'),
            year: new identifier_node_1.IdentifierNode('number'),
        };
    }
}
exports.MysqlAdapter = MysqlAdapter;
