// kysely-codegen/dist/constants/definitions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JSON_VALUE_DEFINITION = exports.JSON_PRIMITIVE_DEFINITION = exports.JSON_OBJECT_DEFINITION = exports.JSON_ARRAY_DEFINITION = exports.GLOBAL_DEFINITIONS = void 0;
const nodes_1 = require("../nodes");
const extends_clause_node_1 = require("../nodes/extends-clause-node");
const generic_expression_node_1 = require("../nodes/generic-expression-node");
const identifier_node_1 = require("../nodes/identifier-node");
const infer_clause_node_1 = require("../nodes/infer-clause-node");
const mapped_type_node_1 = require("../nodes/mapped-type-node");
const union_expression_node_1 = require("../nodes/union-expression-node");
exports.GLOBAL_DEFINITIONS = {
    Generated: [
        ['T'],
        new extends_clause_node_1.ExtendsClauseNode('T', new generic_expression_node_1.GenericExpressionNode('ColumnType', [
            new infer_clause_node_1.InferClauseNode('S'),
            new infer_clause_node_1.InferClauseNode('I'),
            new infer_clause_node_1.InferClauseNode('U'),
        ]), new generic_expression_node_1.GenericExpressionNode('ColumnType', [
            new identifier_node_1.IdentifierNode('S'),
            new union_expression_node_1.UnionExpressionNode([
                new identifier_node_1.IdentifierNode('I'),
                new identifier_node_1.IdentifierNode('undefined'),
            ]),
            new identifier_node_1.IdentifierNode('U'),
        ]), new generic_expression_node_1.GenericExpressionNode('ColumnType', [
            new identifier_node_1.IdentifierNode('T'),
            new union_expression_node_1.UnionExpressionNode([
                new identifier_node_1.IdentifierNode('T'),
                new identifier_node_1.IdentifierNode('undefined'),
            ]),
            new identifier_node_1.IdentifierNode('T'),
        ])),
    ],
};
exports.JSON_ARRAY_DEFINITION = new nodes_1.ArrayExpressionNode(new identifier_node_1.IdentifierNode('JsonValue'));
exports.JSON_OBJECT_DEFINITION = new mapped_type_node_1.MappedTypeNode(new identifier_node_1.IdentifierNode('JsonValue'));
exports.JSON_PRIMITIVE_DEFINITION = new union_expression_node_1.UnionExpressionNode([
    new identifier_node_1.IdentifierNode('boolean'),
    new identifier_node_1.IdentifierNode('null'),
    new identifier_node_1.IdentifierNode('number'),
    new identifier_node_1.IdentifierNode('string'),
]);
exports.JSON_VALUE_DEFINITION = new union_expression_node_1.UnionExpressionNode([
    new identifier_node_1.IdentifierNode('JsonArray'),
    new identifier_node_1.IdentifierNode('JsonObject'),
    new identifier_node_1.IdentifierNode('JsonPrimitive'),
]);
