// @pothos/core/lib/types/type-params.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    outputShapeKey: function() {
        return outputShapeKey;
    },
    parentShapeKey: function() {
        return parentShapeKey;
    },
    abstractReturnShapeKey: function() {
        return abstractReturnShapeKey;
    },
    inputShapeKey: function() {
        return inputShapeKey;
    },
    inputFieldShapeKey: function() {
        return inputFieldShapeKey;
    },
    outputFieldShapeKey: function() {
        return outputFieldShapeKey;
    },
    typeBrandKey: function() {
        return typeBrandKey;
    }
});
const outputShapeKey = Symbol.for('Pothos.outputShapeKey');
const parentShapeKey = Symbol.for('Pothos.parentShapeKey');
const abstractReturnShapeKey = Symbol.for('Pothos.abstractReturnShapeKey');
const inputShapeKey = Symbol.for('Pothos.inputShapeKey');
const inputFieldShapeKey = Symbol.for('Pothos.inputFieldShapeKey');
const outputFieldShapeKey = Symbol.for('Pothos.outputFieldShapeKey');
const typeBrandKey = Symbol.for('Pothos.typeBrandKey');
