// @serverless-stack/resources/dist/ViteStaticSite.js
import path from "path";
import fs from "fs-extra";
import { StaticSite } from "./StaticSite.js";
/////////////////////
// Construct
/////////////////////
/**
 * The `ViteStaticSite` construct is a higher level CDK construct that makes it easy to create a Vite single page app.
 *
 * @example
 *
 * Deploys a Vite app in the `path/to/src` directory.
 *
 * ```js
 * new ViteStaticSite(stack, "Site", {
 *   path: "path/to/src",
 * });
 * ```
 */
export class ViteStaticSite extends StaticSite {
    constructor(scope, id, props) {
        const { path: sitePath, environment, typesPath } = props || {};
        // Show warning
        const app = scope.node.root;
        app.reportWarning("usingViteStaticSite");
        // generate buildCommand
        let defaultBuildCommand = "npm run build";
        if (fs.existsSync(path.join(sitePath, "yarn.lock"))) {
            defaultBuildCommand = "yarn build";
        }
        super(scope, id, {
            indexPage: "index.html",
            errorPage: "redirect_to_index_page",
            buildCommand: defaultBuildCommand,
            buildOutput: "dist",
            fileOptions: [
                {
                    exclude: "*",
                    include: "*.html",
                    cacheControl: "max-age=0,no-cache,no-store,must-revalidate",
                },
                {
                    exclude: "*",
                    include: ["*.js", "*.css"],
                    cacheControl: "max-age=31536000,public,immutable",
                },
            ],
            vite: {
                types: typesPath,
            },
            ...props,
        });
    }
}
